pkg/metallb/frr.go

```diff
@@ -0,0 +1,559 @@
+		builder.Definition.Name, builder.Definition.Namespace,
+	)
+
+	if !builder.Exists() {
+		return builder, fmt.Errorf("FrrConfiguration cannot be deleted because it does not exist")
+	}
+
+	err := builder.apiClient.Resource(
+		GetFrrConfigurationGVR()).Namespace(builder.Definition.Namespace).Delete(
+		context.TODO(), builder.Definition.Name, metav1.DeleteOptions{})
+
+	if err != nil {
+		return builder, fmt.Errorf("can not delete FRRConfiguration: %w", err)
+	}
+
+	builder.Object = nil
+
+	return builder, nil
+}
+
+// GetFrrConfigurationGVR returns FrrConfiguration GroupVersionResource which could be used for Clean function.
+func GetFrrConfigurationGVR() schema.GroupVersionResource {
+	return schema.GroupVersionResource{
+		Group: FRRAPIGroup, Version: APIVersion, Resource: "frrconfigurations",
+	}
+}
+
+// Exists checks whether the given FrrConfiguration exists.
+func (builder *FrrConfigurationBuilder) Exists() bool {
+	if valid, _ := builder.validate(); !valid {
+		return false
+	}
+
+	glog.V(100).Infof(
+		"Checking if FrrConfiguration %s exists in namespace %s",
+		builder.Definition.Name, builder.Definition.Namespace)
+
+	var err error
+	builder.Object, err = builder.Get()
+
+	return err == nil || !k8serrors.IsNotFound(err)
+}
+
+// WithBGPRouter defines a BGP router as localASN.
+func (builder *FrrConfigurationBuilder) WithBGPRouter(localASN uint32) *FrrConfigurationBuilder {
+	if valid, _ := builder.validate(); !valid {
+		return builder
+	}
+
+	glog.V(100).Infof(
+		"Define BGP router with an ASN: %v", localASN)
+
+	if strconv.Itoa(int(localASN)) == "" {
+		glog.V(100).Infof("The local ASN of the frrConfiguration can not be empty")
+
+		builder.errorMsg = "frrConfiguration 'local AS' cannot be empty"
+	}
+
+	if builder.errorMsg != "" {
+		return builder
+	}
+
+	if len(builder.Definition.Spec.BGP.Routers) == 0 {
+		builder.Definition.Spec.BGP.Routers = []frrtypes.Router{{ASN: localASN}}
+	} else {
+		builder.Definition.Spec.BGP.Routers = append(builder.Definition.Spec.BGP.Routers, frrtypes.Router{ASN: localASN})
+	}
+
+	return builder
+}
+
+// WithToReceiveModeAll allows all incoming routes.
+func (builder *FrrConfigurationBuilder) WithToReceiveModeAll(bgpPeerIP string,
+	remoteAS uint32) *FrrConfigurationBuilder {
+	if valid, _ := builder.validate(); !valid {
+		return builder
+	}
+
+	glog.V(100).Infof(
+		"Allows the FRR nodes to receive all routes in namespace %s with mode type: all",
+		builder.Definition.Namespace)
+
+	if builder.errorMsg != "" {
+		return builder
+	}
+
+	if bgpPeerIP == "" {
+		glog.V(100).Infof("The BGP peer IP should not be blank")
+
+		builder.errorMsg = fmt.Sprintf("the bgpPassword %s is an empty string", bgpPeerIP)
+	}
+
+	if strconv.Itoa(int(remoteAS)) == "" {
+		glog.V(100).Infof("The remoteASN of the frrConfiguration can not be empty")
+
+		builder.errorMsg = "frrConfiguration 'remoteAS' cannot be empty when defining receive all mode"
+	}
+
```

Comment:
move it here:
	if builder.errorMsg != "" {
		return builder
	}

Reply:
moved